extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPuller.java

```diff
@@ -86,11 +88,18 @@ public String getName()
       public InputStream openInputStream() throws IOException
       {
         try {
-          streamAcquired = true;
-          return s3Obj.getDataInputStream();
+          synchronized (inputStreamOpener) {
+            if (streamAcquired) {
+              return storageObject.getDataInputStream();
+            }
+            // lazily promote to full GET
+            streamAcquired = true;
```

Comment:
what should the behavior be if getObject fails? Right now it seems like this will throw an exception the first time, and then return null on subsequent calls.

could be better to put `streamAcquired = true` after getObject succeeds?